config: fix config file load order

oliver-sanders opened the pull request:
* Turns out Jupyter Server config_file_paths are loaded in reverse order.
* Fixes the CYLC_SITE_CONF_PATH env var override (which wasn't adding
  `/hub` to the path.

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [ ] Appropriate tests are included (unit and/or functional).
- [x] No changelog (bug has not been released).
- [x] No documentation update required.
- [x] No dependency changes.

datamel reviewed (commented):
I am still having problems with getting config items to load... It is possible my test is flawed, I will detail it below.

- I have `CYLC_SITE_CONF_PATH` exported to: `~/etc/cylc` in my .bashrc,
- In the file `~/etc/cylc/hub/jupyter_config.py` I have `c.CylcUIServer.scan_interval = 60  # PT60S`
- In line 256 of `app.py`, I have added `print(f".!!!!!!!...scan interval is ...{self.scan_interval}")`
- I then run `cylc hub --debug`.
The output is `".!!!!!!!...scan interval is ...5.0"`, which is the default rather than the 60 I have set in the config file.

Can you reproduce this or spot any flaw in my test?

MetRonnie commented:
I can reproduce that. It looks like no exported env vars (including `CYLC_SITE_CONF_PATH`) are getting through to `app.py`? I put
```
print(f".!!!!!!!...ENV:\n{os.environ}")
```
in `__init__` instead of the scan interval

MetRonnie commented:
I used Mel's suggestion of putting
```python
c.Spawner.environment = {
    'CYLC_SITE_CONF_PATH': f'{os.getenv("CYLC_SITE_CONF_PATH")}'
}
```
in `~/.cylc/hub/jupyter_config.py` and now the `CYLC_SITE_CONF_PATH` is coming through successfully, however `self.scan_interval` is still `5.0`, not `60`

oliver-sanders commented:
Can't replicate that I'm afraid.

`c.Spawner` only applies in hub mode, you should be able to test this in hubless mode which reduces the number of moving parts.

I've added some default config to automatically pass through the relevant `CYLC_*` environment variables and added a debug statement in the CylcUIServer which lists all of the configuration that have been set to make testing a little easier.

Here's how I tested it:

```
.cylc/
├── hub
│   └── jupyter_config.py
└── site_conf
    └── hub
        └── jupyter_config.py
```

`.cylc/hub/jupyter_config.py`:

```python
c.CylcUIServer.scan_interval = 3
```

`.cylc/site_conf/hub/jupyter_config.py`:

```python
c.CylcUIServer.scan_interval = 42
```

```
$ cylc gui --debug
...
  * scan_interval = 3
$ CYLC_SITE_CONF_PATH=$HOME/.cylc/site_conf cylc gui --debug
...
  * scan_interval = 42
```

To test in hub mode I had to patch the `--debug` arg onto the spawner command for some reason:

```
$ cylc hub --debug
# refresh browser tab to spawn UIS
...
  * scan_interval = 3
$ CYLC_SITE_CONF_PATH=$HOME/.cylc/site_conf cylc hub --debug
# refresh browser tab to spawn UIS
...
  * scan_interval = 42
```

oliver-sanders commented:
> Can you reproduce this or spot any flaw in my test?

Ah, yes...

You were adding your print statement into the `__init__` method, however, the config has not yet been loaded at this stage (as the object loads the config for itself) so were always getting the default value.

This was perfectly sensible as the periodic callback was started there using the scan interval config which is a bug 😠.

oliver-sanders commented:
Hopefully that should do the trick.

MetRonnie commented:
This is working for me with `cylc gui --debug` as advertised. However, the `cylc hub --debug`  
>``` 
>$ cylc hub --debug
># refresh browser tab to spawn UIS
>...
>  * scan_interval = 3
>$ CYLC_SITE_CONF_PATH=$HOME/.cylc/site_conf cylc hub --debug
># refresh browser tab to spawn UIS
>...
>  * scan_interval = 42
>```
is not working for me -  I don't see the configuration printed out

oliver-sanders commented:
>  However, the cylc hub --debug

When I tried this the `--debug` context wasn't being carried through the spawner properly for some reason so I appended the `--debug` argument to `c.Spawner.command`.

> To test in hub mode I had to patch the --debug arg onto the spawner command for some reason:

(otherwise just bodge the `log.debug` to `log.info` 😁)

MetRonnie reviewed (approved):
Am happy this is working as expected
